### PR TITLE
New Github workflow that provide links to test-configs but reuse contribution repo for report generation

### PR DESCRIPTION
**Pull request by relentless-pursuit**
Resolves: #15234 

doc how to use: https://github.com/checkstyle/test-configs/tree/main/AbbreviationAsWordInName/Example1#example1-configs

for later: 

https://github.com/checkstyle/test-configs/issues/131
https://github.com/checkstyle/test-configs/issues/139
https://github.com/checkstyle/test-configs/issues/140

example of usage: https://github.com/checkstyle-gsoc/checkstyle/pull/7
https://github.com/checkstyle-gsoc/checkstyle/pull/16
or bunch of any other PR: https://github.com/checkstyle-gsoc/checkstyle/pulls

`also done:`
The code changes also included the rocket symbol not working. This has been fixed and tested. 
image
Link to the PR testing: https://github.com/checkstyle-gsoc/checkstyle/pull/7

`also done:`
We need one PR at a time. 
Screenshot 2024-07-24 at 1 10 24 AM

As you can see with changes in workflow to option c, i can only run the latest. Others will be cancelled in the same PR.

image
Link to PR: https://github.com/checkstyle-gsoc/checkstyle/pull/11

### CODE REVIEW

**Comment by relentless-pursuit** [comment]
image

**Comment by romani** [comment]
Please make CI green meanwhile we do review.

**Comment by romani** [comment]
Share link to PR where you did this testing

**Comment by romani** [comment]
Please confirm by test that only single workflow will trigger on trigger phrase. Either existing workflow or new.

File: .github/workflows/regression-report.yml
Comment (romani) [inline .github/workflows/regression-report.yml:68]: Readme was designed to target user convenience.
This workflow should expect files to be named in certain way.
As described at https://github.com/checkstyle/test-configs/blob/main/.github/workflows/CONTRIBUTING.md#custom-configuration-bundles

File: .github/workflows/regression-report.yml
Comment (romani) [inline .github/workflows/regression-report.yml:70]: Instead of readme, take config.xml directly.

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Comment by relentless-pursuit** [comment]
> Share link to PR where you did this testing

https://github.com/checkstyle-gsoc/checkstyle/pull/2

https://github.com/checkstyle-gsoc/checkstyle/pull/2#issuecomment-2135918870
https://github.com/checkstyle-gsoc/checkstyle/pull/2#issuecomment-2136068027

**Comment by relentless-pursuit** [comment]
https://github.com/checkstyle-gsoc/checkstyle/actions/runs/9914040056/job/27392335628

Here, is the workflow running. 

This is the link to the PR: https://github.com/checkstyle-gsoc/checkstyle/pull/1 where I am testing new changes.

**Comment by relentless-pursuit** [comment]
https://github.com/checkstyle-gsoc/checkstyle/actions/runs/9914040056/job/27392335628

Here, is the workflow running. 

This is the link to the PR: https://github.com/checkstyle-gsoc/checkstyle/pull/1 where I am testing new changes.

File: .github/workflows/regression-report.yml
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: Done

File: .github/workflows/regression-report.yml
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: Done

**Comment by romani** [comment]
here is diff - https://www.diffchecker.com/Oj9yiOZe/ with original action

**Comment by romani** [comment]
@relentless-pursuit , please review all CIs and fix them all. Only one failure is expected is for  nonaproved issue.

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit , please review all CIs and fix them all. Only one failure is expected is for nonaproved issue.

yes, trying to fix them

**Comment by romani** [comment]
Pull latest code from main repository, here are commands https://checkstyle.org/beginning_development.html#Starting_Development , there is video on how to do this

**Comment by relentless-pursuit** [comment]
> Pull latest code from main repository, here are commands https://checkstyle.org/beginning_development.html#Starting_Development , there is video on how to do this

I know that. I was just focused on the lin github issue. Now, i have rebased. I guess other jobs related to checkstyle won't fail.

**Comment by romani** [comment]
https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/50203364/job/6err460sotf7qxm1#L115

Before push please do on your local ` mvn verify` it will save you a lot of time, as feedback on quality will be quicker.

**Review by nrmancuso — DISMISSED** [review]
existing workflow is complicated and fragile, we won't be reusing this.

**Comment by romani** [comment]
We do copy to inherit all functionality, and improve step by step all internals.
We don't want to interfere with existing projects to make problems for them to run regression in next few months. 
At the end we hope to be in state where we will substitute existing workflow. But we will see this at the end of project.

File: .ci/diff-report.sh
Comment (romani) [inline .ci/diff-report.sh:15]: please do not change any existing code
we should not damage existing functionality at all.
Others should continue to use old reports workflow without any disruption.

**Review by romani — CHANGES_REQUESTED** [review]
item:

File: .ci/diff-report.sh
Comment (relentless-pursuit) [reply .ci/diff-report.sh:15]: > please do not change any existing code we should not damage existing functionality at all. Others should continue to use old reports workflow without any disruption.

but then the CI was failing saying the workflow doesn't require such variables. Are we also trying to keep the same functionality in this workflow. I mean in this workflow, i am not parsing PR section, or anything else, hence, no use of default project links.

Comment (romani) [reply .ci/diff-report.sh:15]: please share link to violation.

we should not affect existing functionality, we can copy files if we need to modify something.

Comment (relentless-pursuit) [reply .ci/diff-report.sh]: Recent test links:
https://github.com/checkstyle-gsoc/checkstyle/actions/runs/9923541582/job/27413791112

https://github.com/checkstyle-gsoc/checkstyle/pull/1

File: .ci/diff-report.sh
Comment (romani) [inline .ci/diff-report.sh:43]: not ok.

Lets create separate PR and just refactor/rename LINK_FROM_PR to  DIFF_PROJECTS_LINK.
before execution of this shell scripts, github action will do 
`DIFF_PROJECTS_LINK: ${{ needs.parse_body.outputs.projects_link }}`
at
https://github.com/checkstyle/checkstyle/blob/6d8f8f2f9064ab7706fd5c1dd5c715fe0b4ac599/.github/workflows/diff-report.yml#L117-L120

after  merge of such simple refactoring
new action will do `DIFF_PROJECTS_LINK: ${{ needs.parse_comment.outputs.projects_link }}` before action execution.
So no updates in existing will be required.

**Review by romani — CHANGES_REQUESTED** [review]
items:

**Comment by romani** [comment]
@relentless-pursuit , please make web diff link of new workflow and existing and make comment  on why code is different.
last time I noticed you removed concurency was not present, please make it.

and step "name: Prepare message" was very different, please explain.

**Comment by relentless-pursuit** [comment]
> @relentless-pursuit , please make web diff link of new workflow and existing and make comment on why code is different. last time I noticed you removed concurency was not present, please make it.
> 
> and step "name: Prepare message" was very different, please explain.

https://www.diffchecker.com/HvHGkduL/

This is the new web diff. And, i had removed prepare message after consulting with Romani. Earlier, we aimed to post back a response to comment section with a link to the job being run in case the job fails. Now, it doesn't do that and mimicks the earlier function.

File: .github/workflows/regression-report.yml
Comment (romani) [inline .github/workflows/regression-report.yml:39]: combine to single variable - `config_bundle_path` 

this action should not care about module name at all, it just can run anything from test-configs repo. User define in comment path bundle.

File: .github/workflows/regression-report.yml
Comment (romani) [inline .github/workflows/regression-report.yml:90]: set_branch

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .github/workflows/regression-report.yml
Comment (romani) [reply .github/workflows/regression-report.yml:39]: we do not need this variables at all. just remove them

File: .ci/diff-report.sh
Comment (romani) [reply .ci/diff-report.sh:43]: https://github.com/checkstyle/checkstyle/issues/15261 is merged, please rebase, and lets see what else is problem.

File: .github/workflows/regression-report.yml
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: done

File: .github/workflows/regression-report.yml
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: done

**Comment by relentless-pursuit** [comment]
https://github.com/checkstyle-gsoc/checkstyle/actions/runs/9931304825/job/27431022523

new changes runner

File: .github/workflows/regression-report.yml
Comment (romani) [inline .github/workflows/regression-report.yml:35]: or startWith `GitHub, generate report for`

Comment (romani) [reply .github/workflows/regression-report.yml:35]: add space after `for` to both trigger phrases.

File: .github/workflows/regression-report.yml
Comment (romani) [inline .github/workflows/regression-report.yml:65]: `$(echo "$COMMENT_BODY" | sed -E 's/.*, generate report for //'`

it replace phrase to empty string.

File: .github/workflows/regression-report.yml
Comment (romani) [reply .github/workflows/regression-report.yml:35]: and do extra check for it is PR - https://github.com/orgs/community/discussions/25389#discussioncomment-3596273
to be more narrow and better then original workflow.

**Review by romani — CHANGES_REQUESTED** [review]
last items:

File: .github/workflows/regression-report.yml
Comment (romani) [inline .github/workflows/regression-report.yml:5]: PR comment created

File: .github/workflows/regression-report.yml
Comment (romani) [inline .github/workflows/regression-report.yml:22]: Lets remove "edited", I am afraid that it will be too much types of events to handle (we already have long queue for actions) and 
we use "github.run_id" for concurrent, so it will be not clear why some reports generated, as we will not see easily first version of comment.

We can reconsider it later on.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .github/workflows/regression-report.yml
Comment (romani) [reply .github/workflows/regression-report.yml:22]: https://github.com/checkstyle/test-configs/issues/121

Comment (romani) [reply .github/workflows/regression-report.yml:22]: https://github.com/checkstyle/test-configs/issues/122

**Comment by relentless-pursuit** [comment]
Runner for latest changes: https://github.com/checkstyle-gsoc/checkstyle/actions/runs/9948070301/job/27481947080

File: .github/workflows/regression-report.yml
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: > Lets remove "edited", I am afraid that it will be too much types of events to handle (we already have long queue for actions) and we use "github.run_id" for concurrent, so it will be not clear why some reports generated, as we will not see easily first version of comment.
> 
> We can reconsider it later on.

Done

File: .github/workflows/regression-report.yml
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: Done

File: .github/workflows/regression-report.yml
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: Done

File: .github/workflows/regression-report.yml
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: Done

File: .github/workflows/regression-report.yml
Comment (romani) [inline .github/workflows/regression-report.yml:38]: github.event.issue.pull_request
&& (startsWith(github.event.comment.body, 'Github, generate report for ')
    || startsWith(github.event.comment.body, 'GitHub, generate report for '))

**Review by romani — CHANGES_REQUESTED** [review]
Last

**Comment by romani** [comment]
Please fix CI failures

**Comment by relentless-pursuit** [comment]
https://github.com/checkstyle-gsoc/checkstyle/actions/runs/9962679825/job/27527015039

runner for latest code changes

**Comment by relentless-pursuit** [comment]
https://github.com/checkstyle-gsoc/checkstyle/issues/4

Link of the issue section which doesn't trigger the workflow after writing the command intended to trigger the workflow. So, far the workflow is trigger only if it is a PR comment.

File: .github/workflows/regression-report.yml
Comment (romani) [reply .github/workflows/regression-report.yml]: @relentless-pursuit , please place && to front, same for ||

Comment (romani) [reply .github/workflows/regression-report.yml]: @relentless-pursuit , please put a bit more space before ||

To make it visually clear on priority of execution of such operations.

Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml:36]: done

**Comment by relentless-pursuit** [comment]
image

image


the above screenshots show that making a comment on the issue doesn't trigger any workflow. it is only reserved for PR comment.

**Review by romani — APPROVED** [review]
Ok to merge.

All other improvements for implementation will done in next PRs.
This works and others can start to use it they see it convenient.

Merge of this will also unblock testing of generated configs one by one

**Comment by rnveach** [comment]
https://github.com/checkstyle/checkstyle/pull/15235#issuecomment-2224107611
So we get a random error message if user mis-spells check or example? What if user uses `\` instead of `/`? What if only check if given? Is the workflow sensitive to white space or new line after?

**Comment by romani** [comment]
User friendly messages are out of scope for now, this is just basic implementation that is already usable. Very fragile on user typos.

There will be a way more improvements but I highly recommend to do this in separate PRs.

**Comment by romani** [comment]
@rnveach , please review, we need to merge base version to start extensive testing and improvement on inner parts.

**Comment by romani** [comment]
Another example. 
At https://github.com/checkstyle/checkstyle/pull/15334 , instead of placing links to PR.
@mahfouz72  can just do `Github, generate report for NeedBraces/all-examples-in-one`

**Comment by rnveach** [comment]
> There will be a way more improvements but I highly recommend to do this in separate PRs.

Then point to issue where it is documented current issues and what will be fixed in the future.

I expected to see documentation of issues, especially around things I mentioned above.

**Comment by romani** [comment]
I placed few more links to PR description to mention that it is far from final version, there are a lot of rooms to improve.

**Comment by rnveach** [comment]
@romani 
https://github.com/checkstyle/checkstyle/issues/15234#issuecomment-2253464860

You may merge this if you really want to. I will be abstaining from approving.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle/issues/15380 here is reminder to document, assignment to me

**Pull request merged**
